src/OpenTK/Platform/MacOS/CocoaNativeWindow.cs

```diff
@@ -653,15 +655,15 @@ public override void ProcessEvents()
 
                             OnKeyDown(key, isARepeat);
 
-                            var s = Cocoa.FromNSString(Cocoa.SendIntPtr(e, selCharacters));
+                            var s = Cocoa.FromNSString(Cocoa.SendIntPtr(e, selCharactersIgnoringModifiers));
                             foreach (var c in s)
                             {
                                 int intVal = (int)c;
                                 if (!Char.IsControl(c) && (intVal < 63232 || intVal > 63235))
                                 {
                                     // For some reason, arrow keys (mapped 63232-63235)
                                     // are seen as non-control characters, so get rid of those.
-                                    OnKeyPress(c);
+                                    OnKeyPress(capslockOn ? char.ToUpper(c) : c);
```

Comment:
1) `OnKeyPress` is intended to map to hardware keys, not characters, so you shouldn't be altering the value passed into `OnKeyPress`.
2) `ToUpper` does not take into account localisation.